src/common/tse.c

```diff
@@ -502,22 +497,25 @@ tse_task_prep_callback(tse_task_t *task)
 static bool
 tse_task_complete_callback(tse_task_t *task)
 {
-	struct tse_task_private	*dtp = tse_task2priv(task);
-	uint32_t		 gen, new_gen;
-	struct tse_task_cb	*dtc;
-	struct tse_task_cb	*tmp;
+	struct tse_task_private		*dtp = tse_task2priv(task);
+	struct tse_sched_private        *dsp    = dtp->dtp_sched;
+	uint32_t		 	gen, new_gen;
+	struct tse_task_cb		*dtc;
+	struct tse_task_cb		*tmp;
 
 	/* Take one extra ref-count here and decref before exit, as in dtc_cb() it possibly
 	 * re-init the task that may be completed immediately.
 	 */
-	tse_task_addref(task);
+	tse_task_addref_locked(dtp);
 
 	d_list_for_each_entry_safe(dtc, tmp, &dtp->dtp_comp_cb_list, dtc_list) {
 		int ret;
 
 		d_list_del(&dtc->dtc_list);
 		gen = dtp_generation_get(dtp);
+		D_MUTEX_UNLOCK(&dsp->dsp_lock);
```

Comment:
looks unnecessary to lock when calling tse_task_complete_callback() and unlock here,
now there are two callers of tse_task_complete_callback() -
1. in tse_task_post_process(), the lock inside tse_task_post_process() can ensure the task's tse_task_complete_callback() only be called for one thread and one time
2.  by tse_task_complete(), it is required for user that cannot call tse_task_complete() multiple times for one task.

Even if you do the lock as in this PR, it cannot provide more protection if above 2 cases cannot be ensured.

Or you mean change the lock like this PR can avoid some ThreadSanitizer warning? do you have the warning logs?

Reply:
i believe this is one of the issue it tried to fix for accessing dt_result both here in this function:
https://github.com/daos-stack/daos/blob/master/src/common/tse.c#L522
and in schedule_with_delay (instant == true):
https://github.com/daos-stack/daos/blob/master/src/common/tse.c#L1060
this might be a false positive, but im trying to get rid of all warnings.

Reply:
It would indeed be interesting to see what ThreadSanitizer reported directly.

Reply:
ah i was able to reproduce it. so if we remove the lock here and the new one in sched with delay, we get:
```
WARNING: ThreadSanitizer: data race (pid=3275493)
  Read of size 4 at 0x7b600001e400 by thread T2:
    #0 tse_task_schedule_with_delay src/common/tse.c:1048 (libdaos_common.so+0x896dd)
    #1 tse_task_schedule src/common/tse.c:1058 (libdaos_common.so+0x89dbf)
    #2 shard_task_sched src/object/cli_obj.c:2787 (libdaos.so.2+0x101b61)
    #3 tse_task_list_traverse_adv src/common/tse.c:1334 (libdaos_common.so+0x8ca04)
    #4 obj_shard_task_sched src/object/cli_obj.c:2805 (libdaos.so.2+0xfdc5e)
    #5 dc_obj_query_key src/object/cli_obj.c:6999 (libdaos.so.2+0x13b1ef)
    #6 tse_task_schedule_with_delay src/common/tse.c:1045 (libdaos_common.so+0x896bf)
    #7 tse_task_schedule src/common/tse.c:1058 (libdaos_common.so+0x89dbf)
    #8 dc_array_get_size src/client/array/dc_array.c:1945 (libdaos.so.2+0x219d23)
    #9 tse_task_schedule_with_delay src/common/tse.c:1045 (libdaos_common.so+0x896bf)
    #10 tse_task_schedule src/common/tse.c:1058 (libdaos_common.so+0x89dbf)
    #11 dc_task_schedule src/client/api/task.c:114 (libdaos.so.2+0x4fe1c)
    #12 daos_array_get_size src/client/api/array.c:242 (libdaos.so.2+0x263ff)
    #13 lookup_rel_path src/client/dfs/dfs.c:3633 (libdfs.so+0x110b9)
    #14 dfs_lookup src/client/dfs/dfs.c:3812 (libdfs.so+0x2fee7)
    #15 th_metadata /home/mschaara/source/programs/eq_mth_dfs_md.c:77 (eq_mth_dfs_md+0x40129a)
    #16 <null> <null> (libtsan.so.0+0x2a68d)

  Previous write of size 4 at 0x7b600001e400 by thread T7:
    #0 tse_task_complete_callback src/common/tse.c:516 (libdaos_common.so+0x80a33)
    #1 tse_task_complete src/common/tse.c:859 (libdaos_common.so+0x884e4)
    #2 daos_rpc_cb src/client/api/rpc.c:24 (libdaos.so.2+0x4db58)
    #3 crt_hg_req_send_cb src/cart/crt_hg.c:1364 (libcart.so.4+0x75b19)                                                                                                                                                                     
    #4 hg_core_forward_cb /home/mschaara/source/daos/build/external/debug/mercury/src/mercury.c:990 (libmercury.so.2+0x54bd)
    #5 crt_progress_cond src/cart/crt_context.c:1668 (libcart.so.4+0x1ecc5)
    #6 daos_event_priv_wait src/client/api/event.c:1278 (libdaos.so.2+0x3c6ea)
    #7 dc_task_schedule src/client/api/task.c:124 (libdaos.so.2+0x50044)
    #8 daos_array_get_size src/client/api/array.c:242 (libdaos.so.2+0x263ff)
    #9 lookup_rel_path src/client/dfs/dfs.c:3633 (libdfs.so+0x110b9)
    #10 dfs_lookup src/client/dfs/dfs.c:3812 (libdfs.so+0x2fee7)
    #11 th_metadata /home/mschaara/source/programs/eq_mth_dfs_md.c:77 (eq_mth_dfs_md+0x40129a)
    #12 <null> <null> (libtsan.so.0+0x2a68d)

  Location is heap block of size 1024 at 0x7b600001e400 allocated by thread T2:
    #0 calloc <null> (libtsan.so.0+0x2bfe3)
    #1 d_calloc src/gurt/misc.c:50 (libgurt.so.4+0x22dcc)
    #2 tse_task_create src/common/tse.c:950 (libdaos_common.so+0x88ab8)
    #3 queue_shard_query_key_task src/object/cli_obj.c:6810 (libdaos.so.2+0xfe7cd)
    #4 dc_obj_query_key src/object/cli_obj.c:6982 (libdaos.so.2+0x13ad4d)
    #5 tse_task_schedule_with_delay src/common/tse.c:1045 (libdaos_common.so+0x896bf)
    #6 tse_task_schedule src/common/tse.c:1058 (libdaos_common.so+0x89dbf)
    #7 dc_array_get_size src/client/array/dc_array.c:1945 (libdaos.so.2+0x219d23)
    #8 tse_task_schedule_with_delay src/common/tse.c:1045 (libdaos_common.so+0x896bf)
    #9 tse_task_schedule src/common/tse.c:1058 (libdaos_common.so+0x89dbf)
    #10 dc_task_schedule src/client/api/task.c:114 (libdaos.so.2+0x4fe1c)
    #11 daos_array_get_size src/client/api/array.c:242 (libdaos.so.2+0x263ff)                                                                                                                                                               
    #12 lookup_rel_path src/client/dfs/dfs.c:3633 (libdfs.so+0x110b9)
    #13 dfs_lookup src/client/dfs/dfs.c:3812 (libdfs.so+0x2fee7)
    #14 th_metadata /home/mschaara/source/programs/eq_mth_dfs_md.c:77 (eq_mth_dfs_md+0x40129a)
    #15 <null> <null> (libtsan.so.0+0x2a68d)

  Thread T2 (tid=3275496, running) created by main thread at:
    #0 pthread_create <null> (libtsan.so.0+0x2cd42)
    #1 main /home/mschaara/source/programs/eq_mth_dfs_md.c:173 (eq_mth_dfs_md+0x401617)

  Thread T7 (tid=3275501, running) created by main thread at:
    #0 pthread_create <null> (libtsan.so.0+0x2cd42)
    #1 main /home/mschaara/source/programs/eq_mth_dfs_md.c:173 (eq_mth_dfs_md+0x401617)

SUMMARY: ThreadSanitizer: data race src/common/tse.c:1048 in tse_task_schedule_with_delay
```

Reply:
OK i changed how we set the dt_result in the completion and prep callbacks; and reverted this locking change. the warning is gone now with my latest change.